tree/dataframe/src/RCombinedDS.cxx

```diff
@@ -0,0 +1,195 @@
+// Author: Giulio Eulisse CERN  2/2018
+
+/*************************************************************************
+ * Copyright (C) 1995-2018, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+// clang-format off
+/** \class ROOT::RDF::RCombinedDS
```

Comment:
Is there a more precise name we could use?  Combined is a bit too generic (does not evoke Cartesian product).

Reply:
Actually I should update the documentation, since this now allows any kind of combination between the rows of the two combined datasources, not limiting it to the cross product. The actual logic for how to generate the couples of rows is provided by a `RCombinedDSIndex` derived class. So this is effectively combining the rows of two datasources in a generic way and `RCombinedDS` seems to be a good choice to me. I am of course open to other suggestions. Do you have anything in mind?

Reply:
> since this now allows any kind of combination between the rows

I completely missed that.  So indeed the generality of the name is fine ... except that it 'defaults' silently to CrossProduct.

>  So this is effectively combining the rows of two datasources in a generic way

Since this is the case and the intent (this is a good thing), I would definitively add another example (for example the flat/friend combinations).

And then back to the default, when reading I assumed a default of 'flat/friend' combinations, so I would make that the default .... which then brings in the questions of the interface to change the combination index which currently seems to be of the style:
```
auto combined = std::make_unique<RCombinedDS>(std::move(left), std::move(right), std::make_unique<RCombindedDSCrossJoinIndex>() );
```
which seems a bit verbose, any idea on simplifying it or is that good enough?

Reply:
Ok, I will clarify the documentation, implement the friend / `UNION`-like index and make that the default. Regarding your last question, I am not particularly happy either and I am trying to reduce the boilerplate. To start with I was thinking of having some helper:

```C++
template <typename INDEX>
std::unique_ptr<RCombinedDS> Combine<INDEX>(std::unique_ptr<RDataSource>, std::unique_ptr<RDataSource>, std::string left_prefix = "left_", std::string right_prefix = "right_") {...}
```

but suggestions are welcome.

Another alternative I was thinking about would be to introduce a `template <typename CONCRETE_DATASOURCE> RDataSourceNode` class which removes the need for the `#define protected public` and should allow removing the unique_ptr as well (because the RDataSourceNode would really be treated as a value). So it will become:

```C++
template <typename INDEX, typename LEFT, typename RIGHT>
std::unique_ptr<RCombinedDS> Combine<INDEX>(RDataSourceNode<LEFT>, RDataSourceNode<RIGHT>, std::string left_prefix = "left_", std::string right_prefix = "right_") {...}
```

but I am not sure that will result in less verbose code on the client side...

Reply:
@dpiparo Any suggestions?